File: Demos/08_DFS_demo.py

class TreeNode:
    def __init__(self, value):
        self.val = value
        self.left = None
        self.right = None

def DFS(head, visited):
    if not head:
        return
    
    # Add node to visited list
    visited.append(head.val)
    
    # Recursively call dfs on neighbor subtrees
    DFS(head.left, visited)
    DFS(head.right, visited)

def DFSIterative(head):
    visited = []
    pending = [head] # Stack

    # While there are nodes in the pending stack
    while pending:
        # Fetch node from pending stack
        node = pending.pop()

        # Add node to visited list
        visited.append(node.val)

        # Add neighbors to pending stack
        if node.right: pending.append(node.right)
        if node.left: pending.append(node.left)

    return visited

if __name__ == "__main__":
    tree = TreeNode(1)

    tree.left = TreeNode(2)
    tree.right = TreeNode(3)

    tree.left.left = TreeNode(4)
    tree.left.right = TreeNode(5)
    tree.right.left = TreeNode(6)
    tree.right.right = TreeNode(7)

    expectedOutput = [1, 2, 4, 5, 3, 6, 7]
   
    print("DFS recursive")
    output = []
    DFS(tree, output)
    print(output)
    print(expectedOutput)
    print(output == expectedOutput)
    print()

    print("DFS iterative")
    output = DFSIterative(tree)
    print(output)
    print(expectedOutput)
    print(output == expectedOutput)
    print()

File: Demos/C++/08_DFS_demo.cpp
#include <iostream>
#include <vector>

struct Node
{
    int value;
    Node* left;
    Node* right;

    Node(int value): value(value), left(nullptr), right(nullptr) {};
};

void DFSIterative(Node* root, std::vector<int> &visited_values)
{
    std::vector<Node*> pending = {root};

    while (pending.size() > 0)
    {
        Node* node = pending.back();
        pending.pop_back();

        visited_values.push_back(node->value);

        if (node->right != nullptr)
        {
            pending.push_back(node->right);
        }

        if (node->left != nullptr)
        {
            pending.push_back(node->left);
        }
    }
}

void DFSRecursive(Node* root, std::vector<int> &visited_values)
{
    if (root == nullptr) {
        return;
    }

    visited_values.push_back(root->value);

    DFSRecursive(root->left, visited_values);
    DFSRecursive(root->right, visited_values);
}

void PrintVector(std::vector<int> &nums)
{
    for (int num : nums)
    {
        std::cout << num << " ";
    }
    std::cout << std::endl;
}

int main()
{
    Node* tree;
    
    tree = new Node(1);

    tree->left = new Node(2);
    tree->right = new Node(3);

    tree->left->left = new Node(4);
    tree->left->right = new Node(5);
    tree->right->left = new Node(6);
    tree->right->right = new Node(7);

    std::vector<int> dfs_iterative_output;
    DFSIterative(tree, dfs_iterative_output);
    PrintVector(dfs_iterative_output);

    std::vector<int> dfs_recursive_output;
    DFSRecursive(tree, dfs_recursive_output);
    PrintVector(dfs_iterative_output);

    return 0;
}